src/internal/storage/fileset/merge.go

```diff
@@ -112,24 +114,38 @@ func (mfr *MergeFileReader) Content(ctx context.Context, w io.Writer, opts ...ch
 }
 
 // Hash returns the hash of the file.
+// TODO: It would be good to remove this potential performance footgun, but it would require removing the append functionality.
 func (mfr *MergeFileReader) Hash(ctx context.Context) ([]byte, error) {
-	var resolvedDataRefs []*chunk.DataRef
-	cw := mfr.chunks.NewWriter(ctx, "resolve-writer", func(annotations []*chunk.Annotation) error {
-		if annotations[0].NextDataRef != nil {
-			resolvedDataRefs = append(resolvedDataRefs, annotations[0].NextDataRef)
+	var hashes [][]byte
+	size := index.SizeBytes(mfr.idx)
+	if size >= DefaultBatchThreshold {
+		// TODO: Optimize to handle large files that can mostly be copy by reference?
+		if err := miscutil.WithPipe(func(w io.Writer) error {
+			r := mfr.chunks.NewReader(ctx, mfr.idx.File.DataRefs)
+			return r.Get(w)
+		}, func(r io.Reader) error {
+			uploader := mfr.chunks.NewUploader(ctx, "chunk-uploader-resolver", true, func(_ interface{}, dataRefs []*chunk.DataRef) error {
```

Comment:
@acohen4 We compute the stable hash for a file spread across multiple file sets here and use the `noUpload` option since we are just reading the file.